plugins/linux-pipewire/pipewire-virtualcam.c

```diff
@@ -0,0 +1,318 @@
+	blog(LOG_INFO, "Virtual camera started");
+	struct obs_pipewire_virtualcam_data *obs_pwvc =
+		(struct obs_pipewire_virtualcam_data *)data;
+
+	const struct spa_pod *params[2];
+	uint8_t buffer[1024];
+	struct spa_pod_builder b = SPA_POD_BUILDER_INIT(buffer, sizeof(buffer));
+
+	if (!obs_pw_create_loop(&obs_pwvc->pw_core,
+				"obs-pipewire-virtualcam-thread-loop")) {
+		blog(LOG_WARNING, "[pipewire]: Failed to create loop");
+		goto error;
+	}
+
+	if (!obs_pw_start_loop(&obs_pwvc->pw_core)) {
+		blog(LOG_WARNING, "[pipewire]: Failed to start loop");
+		goto error;
+	}
+
+	if (!obs_pw_create_context(&obs_pwvc->pw_core, -1, NULL, NULL)) {
+		blog(LOG_WARNING, "[pipewire]: Failed to create context");
+		goto error;
+	}
+
+	int width = obs_output_get_width(obs_pwvc->output);
+	int height = obs_output_get_height(obs_pwvc->output);
+	params[0] = build_format(&b, width, height, SPA_VIDEO_FORMAT_RGBA);
+	params[1] = build_format(&b, width, height, SPA_VIDEO_FORMAT_YUY2);
+	obs_pwvc->pw_stream.type = OBS_PW_STREAM_TYPE_OUTPUT;
+
+	if (!obs_pw_create_stream(
+		    &obs_pwvc->pw_stream, "obs-pipewire-virtualcam",
+		    pw_properties_new(PW_KEY_NODE_DESCRIPTION,
+				      "OBS Virtual Camera", PW_KEY_MEDIA_CLASS,
+				      "Video/Source", PW_KEY_MEDIA_ROLE,
+				      "Camera", NULL),
+		    PW_ID_ANY,
+		    (PW_STREAM_FLAG_DRIVER | PW_STREAM_FLAG_MAP_BUFFERS),
+		    &stream_events, params, 2, obs_pwvc)) {
+		blog(LOG_WARNING, "[pipewire]: Failed to create stream");
+		goto error;
+	}
+	return true;
+
+error:
+	blog(LOG_WARNING, "Failed to start virtual camera");
+	return false;
+}
+
+static void virtualcam_stop(void *data, uint64_t ts)
+{
+	UNUSED_PARAMETER(ts);
+	blog(LOG_INFO, "Virtual camera stopped");
+	struct obs_pipewire_virtualcam_data *obs_pwvc =
+		(struct obs_pipewire_virtualcam_data *)data;
+
+	obs_output_end_data_capture(obs_pwvc->output);
+	// stop pipewire stuff
+	obs_pw_stop_loop(&obs_pwvc->pw_core);
+	obs_pw_destroy_stream(&obs_pwvc->pw_stream);
+
+	obs_pw_destroy_context(&obs_pwvc->pw_core);
+	obs_pw_destroy_loop(&obs_pwvc->pw_core);
+}
+
+static void virtual_video(void *data, struct video_data *frame)
+{
+	struct obs_pipewire_virtualcam_data *obs_pwvc =
+		(struct obs_pipewire_virtualcam_data *)data;
+
+	// check if we have a running pipewire stream
+	if (!obs_pwvc->pw_stream.pw_stream_state) {
+		// blog(LOG_INFO, "No node connected");
+		return;
+	}
+
+	blog(LOG_DEBUG, "exporting frame to pipewire");
+	// get buffer
+	struct pw_buffer *pw_buf;
+	if ((pw_buf = pw_stream_dequeue_buffer(obs_pwvc->pw_stream.stream)) ==
+	    NULL) {
+		blog(LOG_WARNING, "pipewire: out of buffers");
+		return;
+	}
+
+	struct spa_buffer *spa_buf = pw_buf->buffer;
+	struct spa_data *d = spa_buf->datas;
+
+	if (spa_buf->n_datas != obs_pwvc->formatinfo.planes) {
+		blog(LOG_INFO,
+		     "pipewire: number of buffer planes is different");
+		return;
+	}
+	// copy data
+	for (uint i = 0; i < obs_pwvc->formatinfo.planes; i++) {
+		if (d[0].data == NULL) {
```

Comment:
It looks like you forgot `i` here and below.

Reply:
thx, fixed